src/util/async/context/BasicExecutionContext.hpp

```diff
@@ -0,0 +1,334 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of clio: https://github.com/XRPLF/clio
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#pragma once
+
+#include "util/Expected.hpp"
+#include "util/async/Concepts.hpp"
+#include "util/async/Error.hpp"
+#include "util/async/Operation.hpp"
+#include "util/async/Outcome.hpp"
+#include "util/async/context/impl/Cancellation.hpp"
+#include "util/async/context/impl/Execution.hpp"
+#include "util/async/context/impl/Strand.hpp"
+#include "util/async/context/impl/Timer.hpp"
+#include "util/async/context/impl/Utils.hpp"
+#include "util/async/impl/ErrorHandling.hpp"
+
+#include <boost/asio.hpp>
+#include <boost/asio/error.hpp>
+#include <boost/asio/strand.hpp>
+#include <boost/asio/thread_pool.hpp>
+
+#include <chrono>
+#include <cstddef>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace util::async {
+namespace detail {
+
+struct AsioPoolStrandContext {
+    using Executor = boost::asio::strand<boost::asio::thread_pool::executor_type>;
+    using Timer = SteadyTimer<Executor>;
+
+    Executor executor;
+};
+
+struct AsioPoolContext {
+    using Executor = boost::asio::thread_pool;
+    using Timer = SteadyTimer<Executor>;
+    using Strand = AsioPoolStrandContext;
+
+    Strand
+    makeStrand()
+    {
+        return {boost::asio::make_strand(executor)};
+    }
+
+    boost::asio::thread_pool executor;
```

Comment:
Executor

Reply:
"An executor is someone who is responsible for executing, or following through on, an assigned task or duty.".
It makes sense, although it may be better to call it something else, idk.
Any ideas?

Reply:
I mean: Executor executor; I think the name is 👍

Reply:
Ah, I see it now 🙏